trace_decoder/src/decoding.rs

```diff
@@ -226,573 +44,531 @@ struct TrieDeltaApplicationOutput {
+        .map(|txn_info| {
+            let is_initial_payload = txn_idx == 0;
+
+            let current_idx = txn_idx;
+            if !txn_info.meta.is_dummy() {
+                txn_idx += 1;
+            }
 
-        // Dummy payloads do not increment this accumulator.
-        // For actual transactions, it will match their position in the block.
-        let mut txn_idx = 0;
+            process_txn_info(
+                current_idx,
+                is_initial_payload,
+                txn_info,
+                &mut curr_block_tries,
+                &mut extra_data,
+                &other_data,
+            )
+            .context(format!("at transaction index {}", txn_idx))
+        })
+        .collect::<anyhow::Result<Vec<_>>>()
+        .context(format!(
+            "at block num {} with chain id {}",
+            other_data.b_data.b_meta.block_number, other_data.b_data.b_meta.block_chain_id
+        ))?;
+
+    if !withdrawals.is_empty() {
+        add_withdrawals_to_txns(&mut txn_gen_inputs, &mut curr_block_tries, withdrawals)?;
+    }
 
-        let mut txn_gen_inputs = self
-            .txn_info
-            .into_iter()
-            .map(|txn_info| {
-                let is_initial_payload = txn_idx == 0;
+    Ok(txn_gen_inputs)
+}
 
-                let current_idx = txn_idx;
-                if !txn_info.meta.is_dummy() {
-                    txn_idx += 1;
+/// Cancun HF specific: At the start of a block, prior txn execution, we
+/// need to update the storage of the beacon block root contract.
+// See <https://eips.ethereum.org/EIPS/eip-4788>.
+fn update_beacon_block_root_contract_storage(
+    trie_state: &mut PartialTrieState,
+    delta_out: &mut TrieDeltaApplicationOutput,
+    nodes_used: &mut NodesUsedByTxn,
+    block_data: &BlockMetadata,
+) -> anyhow::Result<()> {
+    const HISTORY_BUFFER_LENGTH_MOD: U256 = U256([HISTORY_BUFFER_LENGTH.1, 0, 0, 0]);
+    const ADDRESS: H256 = H256(BEACON_ROOTS_CONTRACT_ADDRESS_HASHED);
+
+    let timestamp_idx = block_data.block_timestamp % HISTORY_BUFFER_LENGTH_MOD;
+    let timestamp = rlp::encode(&block_data.block_timestamp).to_vec();
+
+    let root_idx = timestamp_idx + HISTORY_BUFFER_LENGTH_MOD;
+    let calldata = rlp::encode(&U256::from_big_endian(
+        &block_data.parent_beacon_block_root.0,
+    ))
+    .to_vec();
+
+    let storage_trie = trie_state
+        .storage
+        .get_mut(&ADDRESS)
+        .context(format!("missing account storage trie {:x}", ADDRESS))?;
+
+    let mut slots_nibbles = vec![];
+
+    for (slot, val) in [(timestamp_idx, timestamp), (root_idx, calldata)]
+        .iter()
+        .map(|(k, v)| {
+            (
+                Nibbles::from_h256_be(hash(Nibbles::from_h256_be(H256::from_uint(k)).bytes_be())),
+                v,
+            )
+        })
+    {
+        slots_nibbles.push(slot);
+
+        // If we are writing a zero, then we actually need to perform a delete.
+        match val == &ZERO_STORAGE_SLOT_VAL_RLPED {
```

Comment:
```suggestion
       if val == &ZERO_STORAGE_SLOT_VAL_RLPED {
```

Reply:
This is from `develop` - https://github.com/0xPolygonZero/zk_evm/blob/1a54af9e8466eb5bbadea85f11e2925b1f74f23c/trace_decoder/src/decoding.rs#L343 - I may refactor this out in a following PR :)